nixos/lib/test-driver/test-driver.py

```diff
@@ -0,0 +1,680 @@
+        return message
+
+    def log_line(self, message, attributes):
+        self.xml.startElement('line', attributes)
+        self.xml.characters(message)
+        self.xml.endElement('line')
+
+    def log(self, message, attributes={}):
+        eprint(self.maybe_prefix(message, attributes))
+        self.drain_log_queue()
+        self.log_line(message, attributes)
+
+    def enqueue(self, message):
+        self.queue.put(message)
+
+    def drain_log_queue(self):
+        try:
+            while True:
+                item = self.queue.get_nowait()
+                attributes = {
+                    'machine': item['machine'],
+                    'type': 'serial'
+                    }
+                self.log_line(self.sanitise(item['msg']), attributes)
+        except queue.Empty:
+            pass
+
+    def nest_enter(self, message, attributes={}):
+        eprint(self.maybe_prefix(message, attributes))
+        print(message)
+
+        self.xml.startElement('nest', attrs={})
+        self.xml.startElement('head', attributes)
+        self.xml.characters(message)
+        self.xml.endElement('head')
+
+        tic = time.time()
+        self.drain_log_queue()
+
+        return tic
+
+    def nest_exit(self, tic):
+        self.drain_log_queue()
+        toc = time.time()
+        self.log('({:.2f} seconds)'.format(toc - tic))
+
+        self.xml.endElement('nest')
+
+    class log_nest:
+        def __init__(self, logger, message, attributes={}):
+            self.message = message
+            self.logger = logger
+            self.attributes = attributes
+
+        def __enter__(self):
+            self.tic = self.logger.nest_enter(self.message, self.attributes)
+            return
+
+        def __exit__(self, type, value, traceback):
+            self.logger.nest_exit(self.tic)
+
+    def nested(self, message, attributes={}):
+        return self.log_nest(self, message, attributes)
+
+
+class Machine:
+    def __init__(self, args):
+        if 'name' in args:
+            self.name = args['name']
+        else:
+            try:
+                cmd = args['startCommand']
+                self.name = re.search('run-(.+)-vm$', cmd).group(1)
+            except:
+                self.name = 'machine'
+
+        self.script = args.get('startCommand', self.create_startcommand(args))
+
+        tmp_dir = os.environ.get('TMPDIR', '/tmp')
+
+        def create_dir(name):
+            dir = os.path.join(tmp_dir, name)
+            os.makedirs(dir, mode=0o700, exist_ok=True)
+            return dir
+
+        self.state_dir = create_dir('vm-state-{}'.format(self.name))
+        self.shared_dir = create_dir('xchg-shared')
+
+        self.booted = False
+        self.connected = False
+        self.pid = None
+        self.socket = None
+        self.monitor = None
+        self.logger = args['log']
+        self.allow_reboot = args.get('allowReboot', False)
+
+    def create_startcommand(self, args):
```

Comment:
this can be a static method, it's not acessing `self` at all.

Reply:
@flokli i incorporated some of the suggestions, but i would like to discuss others. Can you maybe send me a message via matrix/irc? I hang out in the nixos channels with the nick "jonge".